framework/src/outputs/PNGOutput.C

```diff
@@ -0,0 +1,228 @@
+  if(scalingMin != 0)
+  {
+    // Shiftvalue will be the same magnitude, but
+    // going in the opposite direction of the scalingMin
+    shiftValue -= scalingMin;
+  }
+
+  // Shift the max.
+  scalingMax += shiftValue;
+
+}
+
+
+void PNGOutput::setRGB(png_byte *rgb, Real selection) {
+  // Using an RGB system with Red as 0 - 255, Green as 256 - 511 and
+  // Blue as 512 - 767 gives us our total colorSpectrum of 0 - 767.
+  const Real colorSpectrumMax = 767;
+  // We need to convert the
+  auto color = (int)(selection * colorSpectrumMax);
+  // Make sure everything is within our colorSpectrum.
+  if(color > colorSpectrumMax)
+    color = colorSpectrumMax;
+  if(color < 0)
+    color = 0;
+  Real magnitude = color % 256;
+
+  // Current color scheme: Blue->Red->Yellow->White
+  // Blue->Red
+  if(color < 256)
+  {
+    rgb[0] = magnitude;
+    rgb[1] = 0;
+    rgb[2] = 255-magnitude;
+  }
+  // Red->Yellow
+  else if(color < 512)
+  {
+    rgb[0] = 255;
+    rgb[1] = magnitude;
+    rgb[2] = 0;
+  }
+  // Yellow->White
+  else
+  {
+    rgb[0] = 255;
+    rgb[1] = 255;
+    rgb[2] = magnitude;
+  }
+}
+
+
+void PNGOutput::output(const ExecFlagType & /*type*/)
+{
+  makeMeshFunc();
+  box = MeshTools::create_bounding_box(*_mesh_ptr);
+  makePNG();
+}
+
+void PNGOutput::makePNG() {
+  // Increment the testStep
+  testStep++;
+
+  // Get the max and min of the BoundingBox
+  Point maxPoint = box.max();
+  Point minPoint = box.min();
+
+  // The the total distance on the x and y axes.
+  Real distx = maxPoint(0) - minPoint(0);
+  Real disty = maxPoint(1) - minPoint(1);
+
+  // Create the filename based on base and the test step number.
+  std::string PNGFile2 = PNGFile;
+  PNGFile2 += std::to_string((int)testStep);
+  PNGFile2 += ".png";
+
+
+  FILE * fp = nullptr;
+  png_structrp pngp = nullptr;
+  png_infop infop = nullptr;
+  png_bytep row = nullptr;
+  Real depth = 8;
+  Real width = distx*resolution;
+  Real height = disty*resolution;
+
+  if(testStepToPNG != -1)
+    if(testStep != testStepToPNG)
+      return;
+
+  fp = fopen(PNGFile2.c_str(), "wb");
+
+  if(!fp)
+  {
+    mooseError("Failed to open the file for the PNG output.");
+    return;
+  }
+
+  pngp = png_create_write_struct(PNG_LIBPNG_VER_STRING, nullptr, nullptr, nullptr);
+  if(!pngp)
+  {
+    mooseError("Failed to make the pointer string for the png.");
+    return;
+  }
+
+  infop = png_create_info_struct(pngp);
+  if(!infop)
+  {
+    mooseError("Failed to make an info pointer for the png.");
+    return;
+  }
+
+  png_init_io(pngp, fp);
+
+  // Set up the PNG header.
+  png_set_IHDR (pngp,
+                infop,
+                width,
+                height,
+                depth,
+                (inColor?PNG_COLOR_TYPE_RGB:PNG_COLOR_TYPE_GRAY),
+                PNG_INTERLACE_NONE,
+                PNG_COMPRESSION_TYPE_DEFAULT,
+                PNG_FILTER_TYPE_DEFAULT);
+
+  png_write_info(pngp, infop);
+
+  // Allocate resources.
+  row = new png_byte[(width * 3)+1];
```

Comment:
We are really moving away from raw memory management in MOOSE. This is all local to the routine so it's not a big deal but there are a couple other arguably better ways to handle this. 

1) Stash your raw array allocation in a std::unique_ptr() - there are methods for this.
2) Use a std::vector and the std::vector::data method to get back a pointer to the raw vector.

Reply:
Ok, so I couldn't quite get it to work with the uniqu_ptr, but I think I got it to work with the vector.